src/Juvix/Backends/ArithmeticCircuit/Compilation/Types.hs

```diff
@@ -1,19 +1,183 @@
+import qualified Data.Map as Map
+import Numeric.Natural()
+
+import Protolude
+
+data PrimVal
+  = Element Par.F
+  | Boolean Bool
+  | FEInteger Int
+  | BinOp BinOp Term Term
+  | Op Op Term
+  | If Term Term Term
+  deriving (Show, Eq)
+
+data BinOp
+  = Add
+  | Mul
+  | Sub
+  | Exp
+  | Eq
+  | And
+  | Or
+  deriving (Show, Eq)
+
+data Op = Neg
+  deriving (Show, Eq)
+
+type Term = CoreErased.AnnTerm () PrimVal
+type Type = CoreErased.Type () PrimVal
+
+data CompilationError
+  = NotYetImplemented
+  | SomethingWentWrongSorry
+  | VariableOutOfScope
+  deriving (Eq, Show)
+
+data ArithExpression f
+  = BoolExp (Expr.Expr Circuit.Wire f Bool)
+  | FExp (Expr.Expr Circuit.Wire f f)
+  | NoExp
+
+data Memory = Mem (Map.Map J.Symbol (Int, ArithExpression Par.F)) Int
+
+instance Semigroup Memory where
+  (Mem map_ n) <> (Mem map__ m) = Mem (map_ <> Map.map (shift n) map__) (n + m)
+    where shift n (m, exp) = (n + m, exp)
+
+instance Monoid Memory where
+  mempty = Mem Map.empty 0
+  mappend = (<>)
+
+insert :: J.Symbol -> ArithExpression Par.F -> Memory -> Memory
+insert sy x (Mem map_ n) = Mem (Map.insert sy (n, x) map_) (n + 1)
+
+lookup :: J.Symbol -> Memory -> Maybe (Int, ArithExpression Par.F)
+lookup sy (Mem map_ _) =
+  case Map.lookup sy map_ of
+    Just res -> Just res
+    Nothing -> Nothing
+
+transTerm :: Memory
+     -> Term
+     -> (Memory, ArithExpression Par.F)
+transTerm mem CoreErased.Ann{ CoreErased.term = CoreErased.Prim term } =
+  (mem, transPrim term mem)
+transTerm mem CoreErased.Ann{ CoreErased.term = CoreErased.Var var } =
+  case lookup var mem of
+    Just (n, NoExp) -> (mem, FExp $ input n)
+    Just (_, res) -> (mem, res)
+    Nothing -> panic $ show VariableOutOfScope
+transTerm mem CoreErased.Ann{ CoreErased.term = CoreErased.LamM{ CoreErased.body = body
+                                                          , CoreErased.arguments = arguments}} =
+  let mem' = freshVars mem arguments
+  in transTerm mem' body
+transTerm mem CoreErased.Ann{ CoreErased.term = CoreErased.AppM f params } =
+  case f of
+    CoreErased.Ann { CoreErased.term = CoreErased.LamM { CoreErased.body =  body
+                                                       , CoreErased.arguments = arguments
+                                                       }
+                   } -> let transTermParams = snd . sequenceA $ map (transTerm mem) params
+                            matchedArgParams = zip arguments transTermParams
+                            mem' = foldr (\(sy, val) m -> insert sy val m) mem matchedArgParams
+                         in transTerm mem' body
+    _ -> panic $ show SomethingWentWrongSorry
+
+transPrim :: PrimVal -> Memory -> ArithExpression Par.F
+transPrim (Element f) _ = FExp $ Lang.c f
+transPrim (Boolean b) _ = BoolExp $ Circuit.EConstBool b
+transPrim (FEInteger i) _ = FExp $ Lang.c (fromIntegral i)
+-- I miss lenses
+transPrim (BinOp Add prim prim') mem = let (_, FExp prim1) = transTerm mem prim
```

Comment:
Do we want to ignore the memory updates here? Won't we lose variable mappings?

Also, I think this would be cleaner in `MonadState Memory`